File: frontend/src/app/features/wallet/services/wallet.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../../environments/environment';
import { ApiResponse } from '../../../core/models/api-response.model';
import { WalletTransactionRequest, WalletResponse } from '../models/wallet.model';

@Injectable({ providedIn: 'root' })
export class WalletService {
  private http = inject(HttpClient);
  private base = environment.apiBaseUrl;

  credit(userId: number, req: WalletTransactionRequest): Observable<ApiResponse<WalletResponse>> {
    return this.http.post<ApiResponse<WalletResponse>>(`${this.base}/api/v1/wallets/${userId}/credit`, req);
  }
  debit(userId: number, req: WalletTransactionRequest): Observable<ApiResponse<WalletResponse>> {
    return this.http.post<ApiResponse<WalletResponse>>(`${this.base}/api/v1/wallets/${userId}/debit`, req);
  }
  getBalance(userId: number): Observable<ApiResponse<WalletResponse>> {
    return this.http.get<ApiResponse<WalletResponse>>(`${this.base}/api/v1/wallets/${userId}/balance`);
  }
}

File: backend/digital-gold-wallet/src/main/java/com/goldwallet/digitalgoldwallet/modules/wallet/service/WalletService.java
package com.goldwallet.digitalgoldwallet.modules.wallet.service;

import com.goldwallet.digitalgoldwallet.modules.wallet.dto.request.WalletTransactionRequest;
import com.goldwallet.digitalgoldwallet.modules.wallet.dto.response.WalletResponse;

public interface WalletService {
    WalletResponse credit(Long userId, WalletTransactionRequest request);
    WalletResponse debit(Long userId, WalletTransactionRequest request);
    WalletResponse getBalance(Long userId);
}
